src/mbgl/text/placement.cpp

```diff
@@ -59,6 +59,93 @@ const CollisionGroups::CollisionGroup& CollisionGroups::get(const std::string& s
+        return getLayout().get<style::TextVariableAnchor>();
+    }
+
+    float pixelsToTileUnits;
+    float scale;
+    float pixelRatio;
+
+    bool rotateTextWithMap = getLayout().get<TextRotationAlignment>() == AlignmentType::Map;
+    bool pitchTextWithMap = getLayout().get<TextPitchAlignment>() == AlignmentType::Map;
+    bool rotateIconWithMap = getLayout().get<IconRotationAlignment>() == AlignmentType::Map;
+    bool pitchIconWithMap = getLayout().get<IconPitchAlignment>() == AlignmentType::Map;
+    SymbolPlacementType placementType = getLayout().get<SymbolPlacement>();
+
+    mat4 textLabelPlaneMatrix =
+        getLabelPlaneMatrix(renderTile.get().matrix, pitchTextWithMap, rotateTextWithMap, state, pixelsToTileUnits);
+    mat4 iconLabelPlaneMatrix =
+        (rotateTextWithMap == rotateIconWithMap && pitchTextWithMap == pitchIconWithMap)
+            ? textLabelPlaneMatrix
+            : getLabelPlaneMatrix(
+                  renderTile.get().matrix, pitchIconWithMap, rotateIconWithMap, state, pixelsToTileUnits);
+
+    CollisionGroups::CollisionGroup collisionGroup;
+    ZoomEvaluatedSize partiallyEvaluatedTextSize;
+    ZoomEvaluatedSize partiallyEvaluatedIconSize;
+
+    bool textAllowOverlap = getLayout().get<style::TextAllowOverlap>();
+    bool iconAllowOverlap = getLayout().get<style::IconAllowOverlap>();
+    // This logic is similar to the "defaultOpacityState" logic below in updateBucketOpacities
+    // If we know a symbol is always supposed to show, force it to be marked visible even if
+    // it wasn't placed into the collision index (because some or all of it was outside the range
+    // of the collision grid).
+    // There is a subtle edge case here we're accepting:
+    //  Symbol A has text-allow-overlap: true, icon-allow-overlap: true, icon-optional: false
+    //  A's icon is outside the grid, so doesn't get placed
+    //  A's text would be inside grid, but doesn't get placed because of icon-optional: false
+    //  We still show A because of the allow-overlap settings.
+    //  Symbol B has allow-overlap: false, and gets placed where A's text would be
+    //  On panning in, there is a short period when Symbol B and Symbol A will overlap
+    //  This is the reverse of our normal policy of "fade in on pan", but should look like any other
+    //  collision and hopefully not be too noticeable.
+    // See https://github.com/mapbox/mapbox-gl-native/issues/12683
+    bool alwaysShowText =
+        textAllowOverlap && (iconAllowOverlap || !(getBucket().hasIconData() || getBucket().hasSdfIconData()) ||
+                             getLayout().get<style::IconOptional>());
+    bool alwaysShowIcon =
+        iconAllowOverlap && (textAllowOverlap || !getBucket().hasTextData() || getLayout().get<style::TextOptional>());
+
+    bool hasIconTextFit = getLayout().get<IconTextFit>() != IconTextFitType::None;
+
+    optional<CollisionBoundaries> avoidEdges;
```

Comment:
These properties with public access should be all `const`, no?

Reply:
I need to keep them assignable for stable_sort (and I did not want to complicate this structure)